sdklab/stress_fault/src/stress_fault.ts

```diff
@@ -0,0 +1,259 @@
+const FAULT_INTERVAL_MIN = 2 * SECONDS;
+const FAULT_INTERVAL_MAX = 10 * SECONDS;
+
+// How long iothub delays between receiving the fault command and the actual fault.
+const FAULT_DELAY_SECONDS = 1;
+
+// Transport to use
+let protocol = utils.getTransport('amqp');
+
+// Which faults to inject
+let faults = amqpFaults;
+
+// Awaitable sleep helper
+function sleep(interval: number): Promise<void> {
+  return new Promise((resolve) => setTimeout(resolve, interval));
+}
+
+// Running counter of sent messages
+let globalSendCount = 0;
+
+// Interface for our top-level tasks. A top-level task is an awaitable function that
+// runs for the length of the test.
+interface TopLevelTask {
+  done: boolean;
+  stats: any;
+  objectName: string;
+  run(): Promise<void>;
+}
+
+// Top-level task for a device instance
+class Device implements TopLevelTask {
+  _sendIndex: number = 0;
+  _faultCount: number = 0;
+  done: boolean = false;
+  client: Client;
+  deviceId: string;
+  stats: any = {
+    sendEventCount: 0,
+    totalFaultCount: 0,
+  };
+  objectName: string;
+
+  // Main loop for the top-level task
+  public async run(): Promise<void> {
+    // Create a device instane
+    let description = await deviceIdentityHelper.createDeviceWithSymmetricKey();
+    this.deviceId = description.deviceId;
+    this.objectName = description.deviceId;
+
+    let tasks = [];
+    try {
+      // Create the client and connect.
+      this.client = Client.fromConnectionString(description.connectionString, protocol);
+      await this.client.open();
+
+      // Spin up tasks for operations to run at various intervals.
+      tasks = [
+        this._runAtInterval(this._send.bind(this), D2C_SEND_INTERVAL),
+        this._runAtRandomInterval(this._fault.bind(this), FAULT_INTERVAL_MIN, FAULT_INTERVAL_MAX),
+      ];
+
+      // Wait for one of the sub-tasks to fail or complete.
+      await Promise.race(tasks);
+    } catch (e) {
+      debugErrors(`Exception: ${e}`);
+      throw e;
+    } finally {
+      debug('One device task exited.  Stoping device');
+      this.done = true;
+      try {
+        if (tasks) {
+          await (Promise as any).allSettled(tasks);
+        }
+      } finally {
+        debug('finally');
+        await this.client.close();
+        await deviceIdentityHelper.deleteDevice(this.deviceId);
+        debug('device deleted');
+      }
+    }
+  }
+
+  // Helper function run run an async function at a regular interval.
+  private async _runAtInterval(func: Function, interval: number): Promise<void> {
+    while (!this.done) {
+      await func();
+      await sleep(interval);
+    }
+  }
+
+  // Helper function to run an async function at a random interval.
+  private async _runAtRandomInterval(
+    func: Function,
+    minInterval: number,
+    maxInterval: number
+  ): Promise<void> {
+    while (!this.done) {
+      await func();
+      const interval = Math.floor(Math.random() * (maxInterval - minInterval + 1)) + minInterval;
```

Comment:
Is it possible to seed the random generator with a fixed constant so that we get the same random intervals on every run?

Reply:
The node Math.random generator is not seedable.  Other RNGs exist that are seedable, but I don't want to add that right now.  If we want repeatability later, we can add that.